sys/devfs.c

```diff
@@ -6,94 +6,184 @@
-/* Structure for storage in mnt_data */
-typedef struct devfs_mount { vnode_t *root_vnode; } devfs_mount_t;
-
-static inline devfs_mount_t *devfs_of(mount_t *m) {
-  return (devfs_mount_t *)m->mnt_data;
-}
-
-/* Structure for storing installed devices */
-typedef struct devfs_device {
-  char name[DEVFS_NAME_MAX];
-  vnode_t *dev;
-  TAILQ_ENTRY(devfs_device) list;
-} devfs_device_t;
-
-/* The list of all installed devices */
-typedef TAILQ_HEAD(, devfs_device) devfs_device_list_t;
-static devfs_device_list_t devfs_device_list =
-  TAILQ_HEAD_INITIALIZER(devfs_device_list);
-static mtx_t devfs_device_list_mtx = MUTEX_INITIALIZER(MTX_DEF);
-
-static devfs_device_t *devfs_get_by_name(const char *name) {
-  SCOPED_MTX_LOCK(&devfs_device_list_mtx);
-
-  devfs_device_t *idev;
-  TAILQ_FOREACH (idev, &devfs_device_list, list)
-    if (!strcmp(name, idev->name))
-      return idev;
+#include <dirent.h>
+
+typedef struct devfs_node devfs_node_t;
+typedef TAILQ_HEAD(, devfs_node) devfs_node_list_t;
+
+struct devfs_node {
+  TAILQ_ENTRY(devfs_node) dn_link;
+  char *dn_name;
+  ino_t dn_ino;
+  vnode_t *dn_vnode;
+  devfs_node_t *dn_parent;
+  devfs_node_list_t dn_children;
+};
+
+/* device filesystem state structure */
+typedef struct devfs_mount {
+  mtx_t lock;
+  ino_t next_ino;
+  devfs_node_t root;
+} devfs_mount_t;
+
+static devfs_mount_t devfs = {
+  .lock = MUTEX_INITIALIZER(MTX_RECURSE), .next_ino = 2, .root = {}};
+static vnode_lookup_t devfs_vop_lookup;
+static vnode_readdir_t devfs_vop_readdir;
+static vnodeops_t devfs_vnodeops = {.v_lookup = devfs_vop_lookup,
+                                    .v_readdir = devfs_vop_readdir,
+                                    .v_open = vnode_open_generic};
+
+static devfs_node_t *devfs_find_child(devfs_node_t *parent, const char *name) {
+  SCOPED_MTX_LOCK(&devfs.lock);
+
+  if (parent == NULL)
+    parent = &devfs.root;
+
+  devfs_node_t *dn;
+  TAILQ_FOREACH (dn, &parent->dn_children, dn_link)
+    if (!strcmp(name, dn->dn_name))
+      return dn;
   return NULL;
 }
 
-int devfs_install(const char *name, vnode_t *device) {
-  size_t n = strlen(name);
-  if (n >= DEVFS_NAME_MAX)
-    return -ENAMETOOLONG;
+int devfs_makedev(devfs_node_t *parent, const char *name, vnodeops_t *vops,
+                  void *data) {
+  if (parent == NULL)
+    parent = &devfs.root;
 
-  if (devfs_get_by_name(name) != NULL)
-    return -EEXIST;
+  if (parent->dn_vnode->v_type != V_DIR)
+    return -ENOTDIR;
 
-  devfs_device_t *idev = kmalloc(M_VFS, sizeof(devfs_device_t), M_ZERO);
-  strlcpy(idev->name, name, DEVFS_NAME_MAX);
-  idev->dev = device;
+  devfs_node_t *dn = kmalloc(M_VFS, sizeof(devfs_node_t), M_ZERO);
+  dn->dn_name = kstrndup(M_VFS, name, DEVFS_NAME_MAX);
+  dn->dn_vnode = vnode_new(V_DEV, vops, data);
 
-  WITH_MTX_LOCK (&devfs_device_list_mtx)
-    TAILQ_INSERT_TAIL(&devfs_device_list, idev, list);
+  WITH_MTX_LOCK (&devfs.lock) {
+    dn->dn_ino = ++devfs.next_ino;
+    if (devfs_find_child(parent, name))
```

Comment:
Actually my sugestion was not to skip `WITH_MTX_LOCK`, but to move lines L65-L66 out of `WITH_MTX_LOCK` block.